run-python/python.go

```diff
@@ -0,0 +1,266 @@
+package runner
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"os/exec"
+	"path/filepath"
+
+	"github.com/stoic-cli/stoic-cli-core"
+)
+
+const (
+	getPipCacheKey = "run-python/get-pip.py"
+	getPipURL      = "https://bootstrap.pypa.io/get-pip.py"
+
+	pipCacheBase     = ".pip-cache"
+	readyBase        = ".ready"
+	requirementsBase = ".requirements"
+
+	pipRequirements = "" +
+		"pip==10.0.1 --hash=sha256:717cdffb2833be8409433a93746744b59505f42146e8d37de6c62b430e25d6d7\n" +
+		"setuptools==39.1.0 --hash=sha256:0cb8b8625bfdcc2d43ea4b9cdba0b39b2b7befc04f3088897031082aa16ce186\n" +
+		"virtualenv==15.2.0 --hash=sha256:e8e05d4714a1c51a2f5921e62f547fcb0f713ebbe959e0a7f585cc8bef71d11f\n" +
+		"wheel==0.31.0 --hash=sha256:9cdc8ab2cc9c3c2e2727a4b67c22881dbb0e1c503d592992594c5e131c867107\n" +
+		""
+)
+
+func getPipScript(cache stoic.Cache) (string, error) {
+	var err error
+
+	script, err := ioutil.TempFile("", "get-pip-*.py")
+	if err != nil {
+		return "", fmt.Errorf(
+			"unable to set up temp file for get-pip.py: %v", err)
+	}
+	defer func() {
+		if err != nil {
+			os.Remove(script.Name())
+		}
+		script.Close()
+	}()
+
+	if cacheReader := cache.Get(getPipCacheKey); cacheReader != nil {
+		defer cacheReader.Close()
+
+		_, err = io.Copy(script, cacheReader)
+		if err != nil {
+			return "", fmt.Errorf(
+				"unable to read get-pip.py from cache: %v", err)
+		}
+	} else {
+		resp, err := http.Get(getPipURL)
+		if err != nil {
+			return "", fmt.Errorf(
+				"unable to download get-pip.py script from upstream: %v", err)
+		}
+		defer resp.Body.Close()
+
+		err = cache.Put(getPipCacheKey, io.TeeReader(resp.Body, script))
+		if err != nil {
+			return "", fmt.Errorf(
+				"unable to download get-pip.py script from upstream: %v", err)
+		}
+	}
+
+	return script.Name(), nil
+}
+
+func newPythonEnvironment(root string, python string, cache stoic.Cache) (pythonEnvironment, error) {
```

Comment:
See comment below, here we can return the `interface` instead of `pythonEnvironment`, e.g., `PythonEnvironment`